### PR TITLE
Portable BPF map implementation packages

### PR DESCRIPTION
**Pull request by tommyp1ckles**
Refactoring various map types in order to make type declaration code portable.
Currently, when trying to import/build against map packages (ex. `pkg/maps/lxcmap`) it is not possible to build for non linux platforms.
This is because these packages import pkg/bpf which relies on Linux specific bpf C libraries.

To resolve this, this PR separates struct declarations used for implementing the bpf.MapKey/bpf.MapValue types from their implementation packages. Data struct declarations are moved to `./types` subpackage for each map type (i.e. `pkg/maps/lxcmap/types`).

Implementation is then added by aliasing the type in the original map package. For example:

**note**: code gen headers, such as deepcopy-gen should remain in the implementation package, not with the types type declarations.

`pkg/maps/ipcache/types/ipcache.go`
```go
type Key struct {
	Prefixlen uint32 `align:"lpm_key"`
	Pad1      uint16 `align:"pad1"`
	Pad2      uint8  `align:"pad2"`
	Family    uint8  `align:"family"`
	// represents both IPv6 and IPv4 (in the lowest four bytes)
	IP types.IPv6 `align:"$union0"`
}
```

and `pkg/maps/ipcache/ipcache.go`

```go
// Key implements the bpfTypes.MapKey interface.
// Must be in sync with struct ipcache_key in <bpf/lib/maps.h>
// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf/types.MapKey
type Key ipcacheTypes.Key

func (k Key) NewValue() bpfTypes.MapValue { return &RemoteEndpointInfo{} }

// etc ...
```

To ensure that this constraint remains true, added a CI check that will build all go packages in Cilium named "types" against non Linux GOOS targets to ensure that they are portable (also ensures portability against unix specific libraries such `sys/unix`).

```release-note
Refactor pkg/maps declaration types to be portable
```

### CODE REVIEW

File: .github/workflows/lint-go.yaml
Comment (nbusseneau) [inline .github/workflows/lint-go.yaml:96]: [Nit] Seems like `go-version` has been updated to 1.19.2 in other jobs.

**Review by nbusseneau — APPROVED** [review]
See nit above. Otherwise workflow changes LGTM.

File: pkg/maps/lbmap/types/lbmap.go
Comment (jrajahalme) [inline pkg/maps/lbmap/types/lbmap.go]: Seems like this file is not needed?

File: pkg/maps/lbmap/types/maglev.go
Comment (jrajahalme) [inline pkg/maps/lbmap/types/maglev.go]: Seems like this file is not needed?

File: pkg/maps/lbmap/types/maglev_inner_map.go
Comment (jrajahalme) [inline pkg/maps/lbmap/types/maglev_inner_map.go]: Seems like this file is not needed?

File: pkg/maps/lbmap/types/maglev_outer_map.go
Comment (jrajahalme) [inline pkg/maps/lbmap/types/maglev_outer_map.go]: Seems like this file is not needed?

File: pkg/maps/lbmap/types/types.go
Comment (jrajahalme) [inline pkg/maps/lbmap/types/types.go]: Seems like this file is not needed?

**Review by jrajahalme — APPROVED** [review]
LGTM apart from a few unnecessary files being added?

**Comment by jrajahalme** [comment]
Bunch of Go lint issues:
```
Error: undeclared name: `BPFTestSuite` (typecheck)
  Error: undeclared name: `BPFTestSuite` (typecheck)
  Error: undeclared name: `BPFTestSuite` (typecheck)
  Error: Missed header for check (goheader)
  Error: Missed header for check (goheader)
  Error: type `pad4uint32` is unused (unused)
  Error: func `(*pad4uint32).DeepCopyInto` is unused (unused)
  Error: type `pad2uint8` is unused (unused)
  
  Error: issues found
  Ran golangci-lint in 4847[25](https://github.com/cilium/cilium/actions/runs/3550041134/jobs/5963056538#step:4:27)ms
```

File: pkg/maps/lbmap/types/types.go
Comment (tommyp1ckles) [reply pkg/maps/lbmap/types/types.go]: Woops, yup, this was partially refactored using a program, need to remove no-op cases where it still wrote the file

**Review by joamaki — APPROVED** [review]

File: pkg/maps/README
Comment (pippolo84) [inline pkg/maps/README]: typo: `./lxamp` -> `./lxcmap`

**Review by pippolo84 — APPROVED** [review]
LGTM minus a small typo. Thanks! :rocket:

**Comment by tommyp1ckles** [comment]
/test-1.26-net-next

**Comment by tommyp1ckles** [comment]
/test-1.24-5.4

**Comment by tommyp1ckles** [comment]
@julianwiedmann @borkmann @aditighag I've rebased the PR, this one is prone to conflicts but ill try to resolve any conflicts in next few days to try to get this merged - if I can get the approvals.

**Review by julianwiedmann — COMMENTED** [review]
Looks like datapath was already covered by Jussi - so dropping myself to not hold this up further.

Just two small comments:

- `ci: add lint check to ensure that */types packages are portable.` also touches `pkg/maps/ctmap/`. That part probably belongs into the initial patch?
- should `pkg/bpf/types: fix bpf pkg tests.` also be squashed into the initial patch?

**Pull request closed without merging**
